### ISSUE TITLE
UBSAN issue on ubuntu16_04_clang

### ISSUE THREAD

**Issue by veblush** [GoogleCloudPlatform/layer-definitions#531]
Ubuntu 16.04 with clang image for RBE bazel uses gcc 4.9 libstdc++ (it's strange because Ubuntu 16.04 is bundled with gcc 5.3) This version has the [bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66017) where std::map doesn't handle alignment correctly making UBSAN to report a misaligned address error when testing it. This bug was fixed since gcc 5.2.

gRPC has an [issue](https://github.com/grpc/grpc/pull/19981) with this and I tried to make it to use libc++ instead of libstdc++ but it didn't work (https://github.com/grpc/grpc/pull/20337) because libc++ on the image is only for msan.

Possible solutions would be either of
  1) Backporting the fix from gcc 5. ([patch](https://gcc.gnu.org/viewcvs/gcc?view=revision&revision=228098))
  1) Enabling libc++ for UBSAN, too.
  1) Upgrading the existing libstdc++ to 5.2 or later.
  1) Alternative RBE-Bazel image without this bug?

WDTY?

**Comment by alex1545** [comment]
Thank you for reporting this issue.
I think the best approach to resolve this problem is going with the 3rd option you suggested.
I will report back in this issue once a new Ubuntu 16.04 RBE image is available with this fix.

**Comment by veblush** [comment]
Thank you for taking a look at this. I have questions on this.
- There is a new ubuntu18_04_clang on this repository and when it will be ready to used?
- ubuntu18_04_clang seems to go with libstdc++5 even though Ubuntu 18.04 is bundled with libstdc++7. Is there a reason to use older version of libstdc++? It was a surprise to see ubuntu16_04_clang was not with libstdc++5.

**Comment by alex1545** [comment]
- The RBE Ubuntu18.04 image is currently undergoing the Google Cloud Marketplace Onboarding procedure. This procedure may take up to 2 weeks until this image becomes publicly available. In the meantime, you can build and test the image from this repository.
- There is no particular reason to use an older version of libstdc++ for our images. We will look at the image dependencies once again and update accordingly. And thank you for pointing this out.

**Comment by veblush** [comment]
Thanks for the answer!

**Comment by alex1545** [comment]
Hi @veblush,
We would like to hold on to updating the libstdc++ version in the RBE Ubuntu1604 image because this will cause changes in the toolchain configs that we release.
Are you able to wait for the RBE Ubuntu1804 image to be available? And if so, will that image solve your problem or will you still need the RBE Ubuntu1604 with libstdc++5?

**Comment by veblush** [comment]
It seems acceptable to use the latest version of image instead of the old one. What do you think? @jtattermusch, @nicolasnoble?

**Comment by jtattermusch** [comment]
It's fine to upgrade to newer toolchain if we check that the flakiness rate doesn't increase significantly (or we can file issues for the new failures and commit to fix them quickly).

**Comment by alex1545** [comment]
Thank you for confirming.
We will let you know once the RBE Ubuntu18.04 image becomes publicly available on Cloud Marketplace.

**Comment by alex1545** [comment]
The RBE Ubuntu18.04 image is now public.
It's available at [gcr.io/cloud-marketplace/google/rbe-ubuntu18-04](https://gcr.io/cloud-marketplace/google/rbe-ubuntu18-04).
Please try it out and let us know if you run into any issues.

**Comment by veblush** [comment]
Thank you! I'll let you know how it goes well with gRPC.

**Comment by nlopezgi** [comment]
example target in WORKSPACE to use new container with RBE:

```
load("@bazel_toolchains//rules:environments.bzl", "clang_env")
rbe_autoconfig(
    name = "rbe_ubuntu1804",
    env = clang_env(),
    registry = "gcr.io",
    repository = "cloud-marketplace/google/rbe-ubuntu18-04",
    digest = "sha256:0191ae7277b35b787851168d1900ec36d0a47744315638aa0a00d078ca1dc63c",
)
```

Then you can switch instances of @rbe_default for @rbe_ubuntu1804 in the file pointed to by https://github.com/bazelbuild/bazel-toolchains/blob/master/bazelrc/latest.bazelrc
You should also use the recommended UBSAN flags (https://github.com/grpc/grpc/blob/master/tools/remote_build/rbe_common.bazelrc#L81).
Closing this issue, but please let me know to reopen if you run into issues

**Comment by nlopezgi** [comment]
Actually, I just realized the ubsan toolchain grpc is currently using is still the experimental one in https://github.com/bazelbuild/bazel-toolchains/tree/master/configs/experimental/ubuntu16_04_clang/1.2/bazel_0.28.0/ubsan which will not work with the ubuntu 18.04 container. The process to create this toolchain config is non trivial as it requires manual modification of the cc_toolchain_config.bzl to use clang++ as tool (https://github.com/bazelbuild/bazel-toolchains/blob/master/configs/experimental/ubuntu16_04_clang/1.2/bazel_0.28.0/ubsan/cc_toolchain_config.bzl#L85). Please reach out to me if/when you are interested in setting this up so I can guide you through the process.

**Comment by nlopezgi** [comment]
I just published a toolchain config for bazel 0.29.1 for UBSAN with the latest rbe ubuntu 16.04 container that comes with clang 10.0.0. That PR can serve as a basis for what you need to do to set up an ubsan toolchain for the new rbe-ubuntu-18.04 container

**Comment by veblush** [comment]
Thank for the investigation. I'm now creating [PR](https://github.com/grpc/grpc/pull/20431) to use 18.04 image on gRPC. Please take a look at this and please let me know there is something to be done.
